src/audio/asrc/asrc.c

```diff
@@ -0,0 +1,616 @@
+	if (!dev)
+		return NULL;
+
+	asrc = (struct sof_ipc_comp_asrc *)&dev->comp;
+
+	assert(!memcpy_s(asrc, sizeof(*asrc), ipc_asrc,
+			 sizeof(struct sof_ipc_comp_asrc)));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	/* Get operation mode:
+	 * With eIasPush (0) use fixed input frames count, variable output.
+	 * With eIasPull (1) use fixed output frames count, variable input.
+	 */
+	cd->mode = asrc->operation_mode;
+
+	/* Initialize drift to factor 1 */
+	cd->drift = Q_CONVERT_FLOAT(1.0, 30);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void asrc_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_asrc("asrc_free()");
+
+	rfree(cd->buf);
+	rfree(cd->src_obj);
+	rfree(cd);
+	rfree(dev);
+}
+
+static int asrc_ctrl_cmd(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	trace_asrc_error("asrc_ctrl_cmd()");
+	return -EINVAL;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int asrc_cmd(struct comp_dev *dev, int cmd, void *data,
+		    int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	trace_asrc("asrc_cmd()");
+
+	if (cmd == COMP_CMD_SET_VALUE)
+		ret = asrc_ctrl_cmd(dev, cdata);
+
+	return ret;
+}
+
+static int asrc_trigger(struct comp_dev *dev, int cmd)
+{
+	trace_asrc("asrc_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/* set component audio stream parameters */
+static int asrc_params(struct comp_dev *dev)
+{
+	struct sof_ipc_stream_params *params = &dev->params;
+	struct sof_ipc_comp_asrc *asrc = COMP_GET_IPC(dev, sof_ipc_comp_asrc);
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_asrc("asrc_params()");
+
+	/* Calculate source and sink rates, one rate will come from IPC new
+	 * and the other from params.
+	 */
+	if (asrc->source_rate) {
+		/* params rate is sink rate */
+		cd->source_rate = asrc->source_rate;
+		cd->sink_rate = params->rate;
+		cd->source_frames = ceil_divide(dev->frames * cd->source_rate,
+						cd->sink_rate);
+		cd->sink_frames = dev->frames;
+		/* re-write our params with output rate for next component */
+		params->rate = cd->source_rate;
+	} else {
+		/* params rate is source rate */
+		cd->source_rate = params->rate;
+		cd->sink_rate = asrc->sink_rate;
+		/* re-write our params with output rate for next component */
+		params->rate = cd->sink_rate;
+		cd->source_frames = dev->frames;
+		cd->sink_frames = ceil_divide(dev->frames * cd->sink_rate,
+					      cd->source_rate);
+	}
+
+	/* Use empirical add +10 for target frames number. It's
+	 * slightly more than min. needed and does not increase much peak
+	 * load and buffer memory need too much. The copy() function will
+	 * find the limits and process less frames after buffers levels
+	 * stabilize.
+	 */
+	cd->source_frames_max = cd->source_frames + 10;
+	cd->sink_frames_max = cd->sink_frames + 10;
```

Comment:
why is this needed in the first place? The comment only says that it's harmless but should explain why it was added.

Reply:
Yes, needs explanation. The playback start glitches are avoided with this. It's not clear if it's component fault or pipeline/scheduler fault. The existing SRC also has the SOF issue #1843 issue.